hazelcast/src/main/java/com/hazelcast/map/impl/wan/WanMapEntryView.java

```diff
@@ -69,15 +71,15 @@ public WanMapEntryView(@Nonnull Data dataKey,
                            @Nonnull SerializationService serializationService) {
         checkNotNull(dataKey);
         checkNotNull(serializationService);
-        this.dataKey = dataKey;
-        this.dataValue = dataValue;
+        this.dataKey = serializationService.toDataWithSchema(dataKey);
+        this.dataValue = serializationService.toDataWithSchema(dataValue);
```

Comment:
As I understand, we expect here that the WAN target side (which will deserialize the entry) is on 5.0+ as well. Note for future us: in the final version, we'll need to perform the serialization conditionally based on the negotiation results of the WAN protocol, together with the schema extraction. This also means we will need to bump the WAN protocol version when we make compact serialization format a production feature.

Reply:
Second thought: my above comment assumes `toDataWithSchema()` takes the config into account and it doesn't serialize in the compact format if it's not enabled in the config. @sancar can you confirm this is the case? Otherwise, our WAN compatibility guarantees are broken here.

Reply:
Yes, basically, it will only include the schema in data if
- compact serialization is enabled and the class of the object is registered to be serialized with compact serialization via CompactSerializationConfig
- compact serialization is enabled and the class cannot be serialized with any other mechanism (Portable, Identified, Java serializable, etc.) and the class can be serialized with compact serialization reflectively (zero-config use case)
- this is already a data object and the type of the data is `COMPACT` (hence, it is serialized with a compact serializer before).

For anything else, this method is the same as `toData`

Reply:
> As I understand, we expect here that the WAN target side (which will deserialize the entry) is on 5.0+ as well. Note for future us: in the final version, we'll need to perform the serialization conditionally based on the negotiation results of the WAN protocol, together with the schema extraction. This also means we will need to bump the WAN protocol version when we make compact serialization format a production feature.

I could not follow your comment but let me remind you something first.
5.0 compact format will not work with 5.1 compact. We don't want our users to start using Compact in 5.0 then migrate to 5.1 in production. This will be clearly documented almost everywhere. 

Lets assume that we want to support 5.0 to 5.1 migration for the sake of the argument.
I believe the `trimSchema` on the receiving side should stay forever so that we can support backward compatibility. Note that trimSchema is not an heavy operation if the data does not have the schema. It is just a type-id check then no-op. This way we don't need to change WAN protocol version. 
Does that make sense ?

>Second thought: my above comment assumes toDataWithSchema() takes the config into account and it doesn't serialize in the compact format if it's not enabled in the config. @sancar can you confirm this is the case? Otherwise, our WAN compatibility guarantees are broken here.

Note that toDataWithSchema is noop for any other format. 
It appends the schema to Data if the Data is in the Compact format only. So if it is not enabled in the config, we will not have a Data with compact in the first place.

Reply:
Thanks for the clarification @sancar :+1: 

How WAN is different from client-to-member and member-to-member communications is that we have a much wider compatibility guarantee for WAN. `5.y` is expected to be able to WAN replicate to `4.y` and `5.1` to `5.0` as well. To support this, we added WAN protocol negotiation, which determines the features we can use over a given WAN link. Until now, the serialization format was not in the focus of the WAN protocol negotiation. But now, if the key and/or value types used in an `IMap` are registered for compact format, WAN here will serialize in compact format that the target (if it's `4.y`) will not understand. 

In short, to support WAN compatibility, we should take into account not only the serialization configuration but a - some sort of - inclusion/exclusion set of the formats that can/can't be used _in the given context_. So to keep this mentioned compatibility, it might be desired to use the compact format for "local" (client-to-member, member-to-member) serialization, while we should use an older format for WAN.

This is not an issue as long as compact serialization format is not considered a production feature. But we should solve this before we mark the format production-ready.

I wonder if the same concerns apply to hot restart/persistence. cc @mmedenjak

Reply:
I don't understand this part:
> I believe the trimSchema on the receiving side should stay forever so that we can support backward compatibility.

But what Zoltan says is true and I can try to rephrase if it will make it any clearer. Assuming that the serialised form of the new format isn't indistinguishable from the current formats in 3.x and 4.x, we will have to bump the WAN protocol version at some point.

The reason is as Zoltan says - WAN works across minor and even is meant to work across major versions. Some examples:
• 4.0 should be able to replicate to 4.2
• 4.2 should be able to replicate to 4.0
• 5.0 should be able to replicate to 4.1
• 4.1 should be able to replicate to 5.2
And the way we achieve is - we have a set of supported WAN protocol versions that's exchanged when the connection between clusters is established. If the members can agree on a single version, great, WAN works. If not, the connection is closed and an exception is thrown.

So when compact format is made production-ready (e.g. 5.1), the 5.1 cluster will need to replicate entries to 5.0 clusters (and even 4.2 clusters). If the 5.0 cluster can't deserialise it, it will keep throwing exceptions. The 5.1 cluster needs to be aware that it's connecting to an older cluster and that it either needs to throw an exception to the user - "You're using compact format and you cannot replicate to an older cluster" or it needs to somehow "translate" the compact format into something that the 4.2 cluster will be able to understand (`Portable`?).

Hope that makes sense.

Reply:
I think we should not try to replicate a compact format to a 4.1 cluster. 
Portable is not enough to support all field types of the new format. Even if we manage to hack it somehow, the configuration that user should do on 4.1 cluster will be really confusing. Same object will implement `Portable` in 4.1 but not on 5.1. 
So, I believe we should support only 5.1 onwards for compact and Wan together.

In short, I understand that there are problems and we need to work more closely on Wan when we want to finalize this. I will take a note of this and continue with the pr as is, if that is ok?

Reply:
Thanks for making the note @sancar :+1: 

I think we need to introduce the new WAN protocol version together with 5.1. But I don't know how can we fail-fast here, as the types of the `IMap`/`ICache` will be determined in runtime. It's possible that WAN replication works smoothly for weeks without any issue - because the replicated types are not registered for compact serialization - and then it suddenly fails - after putting an entry that registered for compact serialization. From the user experience point of view, it would be better to fall back to a working serialization format. But I understand it's not trivial (how the serialization should know that it needs to fall back to another format and what that other format should be?).

Reply:
Yes, thanks for adding the note. As you say, since it's not possible to translate the compact format into something an older cluster might understand, we may even have to throw an exception to the user saying that this entry cannot be put into the map because the map is configured to be replicated to an older cluster. And for that, we'll need the WAN protocol version support. Otherwise, everything would seem fine on the source cluster, the entry would be added to the WAN queues for replication after which both the source and target cluster logs would be filled with exceptions since the entry replication is retried indefinitely, until an operator pauses WAN replication and clears the queues.